lib/Sema/CSSyntacticElement.cpp

```diff
@@ -1195,6 +1247,87 @@ bool ConstraintSystem::generateConstraints(AnyFunctionRef fn, BraceStmt *body) {
   return generator.hadError;
 }
 
+bool ConstraintSystem::generateConstraints(SingleValueStmtExpr *E) {
+  auto *S = E->getStmt();
+  auto &ctx = getASTContext();
+
+  auto *loc = getConstraintLocator(E);
+  Type resultTy = createTypeVariable(loc, /*options*/ 0);
+  setType(E, resultTy);
+
+  // Assign contextual types for each of the expression branches.
+  SmallVector<Expr *, 4> scratch;
+  auto branches = E->getSingleExprBranches(scratch);
+  for (auto *branch : branches) {
+    setContextualType(branch, TypeLoc::withoutLoc(resultTy),
+                      CTP_SingleValueStmtBranch);
+  }
+
+  TypeJoinExpr *join = nullptr;
+  if (branches.empty()) {
+    // If we only have statement branches, the expression is typed as Void. This
+    // should only be the case for 'if' and 'switch' statements that must be
+    // expressions that have branches that all end in a throw, and we'll warn
+    // that we've inferred Void.
+    addConstraint(ConstraintKind::Bind, resultTy, ctx.getVoidType(), loc);
+  } else {
+    // Otherwise, we join the result types for each of the branches.
+    join = TypeJoinExpr::forBranchesOfSingleValueStmtExpr(
+        ctx, resultTy, E, AllocationArena::ConstraintSolver);
+  }
+
+  // If this is the single expression body of a closure, we need to account
+  // for the fact that the result type may be bound to Void. This is necessary
+  // to correctly handle the following case:
+  //
+  // func foo<T>(_ fn: () -> T) {}
+  // foo {
+  //   if .random() { 0 } else { "" }
+  // }
+  //
+  // Before if/switch expressions, this was treated as a regular statement,
+  // with the branches being discarded (and we'd warn). We need to ensure we
+  // maintain compatibility by continuing to infer T as Void in the case where
+  // the branches mismatch. This example is contrived, but can occur in the real
+  // world with e.g branches that insert and remove elements from a set, in both
+  // cases the methods have mismatching discardable returns.
+  //
+  // To maintain this behavior, form a disjunction that will attempt to either
+  // bind the expression type to the closure result type, or bind it to Void.
+  // Only if we fail to solve with the closure result type will we attempt with
+  // Void. We can't rely on the usual defaulting of the closure result type,
+  // as we need to solve the conjunction before trying defaults.
+  //
+  // This only needs to happen for cases where the return is implicit, we don't
+  // need to do this with 'return if'. We also don't need to do it for function
+  // decls, as we proactively avoid transforming the if/switch into an
+  // expression if the result is known to be Void.
+  if (auto *CE = dyn_cast<ClosureExpr>(E->getDeclContext())) {
+    if (CE->hasSingleExpressionBody() &&
+        CE->getSingleExpressionBody()->getSemanticsProvidingExpr() == E) {
```

Comment:
We have `hasExplicitResult`, could it be used here?

Reply:
Nice, yup let's use that